app/i18n/translator.html

```diff
@@ -128,111 +165,280 @@
+          selectedFilterVal: filterVal
+        };
+        if (newkey != state.selectedKey) {
+          const previousKeys = [...state.previousKeys];
+          previousKeys?.push({
+            selectedKey: newkey,
+            filterVal: filterVal
+          });
+          delta = { ...delta, previousKeys };
+        }
+        updateState(delta);
         refreshUI();
+        scrollSelectedEntryIntoViewIfNeeded();
       }
 
-      function refreshUI() {
-        const filterVal = document.querySelector('input[name="filter"]:checked').value;
+      function onFilterChange() {
+        dotry(refreshUI);
+        scrollSelectedEntryIntoViewIfNeeded();
+      }
 
+      function scrollSelectedEntryIntoViewIfNeeded() {
+        const el = document.querySelector('li[class="selected"]');
+        if (!el) {
+          return;
+        }
+        const stringList = elID("string_list");
+        const { top, bottom, height } = el.getBoundingClientRect();
+        const stringListRect = stringList.getBoundingClientRect();
+
+        const isScrolledIntoView =
+          top <= stringListRect.top
+            ? stringListRect.top - top <= height
+            : bottom - stringListRect.bottom <= height;
+
+        if (!isScrolledIntoView)
+          el.scrollIntoView({
+            block: "end",
+            inline: "nearest"
+          });
+      }
+
+      function jumpToPreviousKey() {
+        // the last element in state.previousKeys is the selected key
+        const previousKey =
+          state.previousKeys.length > 1 &&
+          state.previousKeys[state.previousKeys.length - 2];
+        if (previousKey) {
+          // remove selected and previous key
+          updateState({
+            previousKeys: [...state.previousKeys].slice(0, -2)
+          });
+          setFilterVal(previousKey.filterVal);
+          changeToKey(previousKey.selectedKey, true);
+        }
+      }
+
+      function refreshUI() {
+        const filterVal = getFilterVal();
         const stringsEl = elID("string_list");
         stringsEl.innerHTML = "";
 
+        if (filterVal === "deleted") {
+          Object.keys(state.deletedEntries).map((k) => {
+            const li = newEl(stringsEl, "li");
+            li.innerText = state.deletedEntries[k];
+            newEl(li, "span").innerText = "";
+
+            if (state.deletedEntries[k] === state.selectedKey) {
+              li.classList.add("selected");
+              selectedEntry = k;
+            }
+
+            li.onclick = () => changeToKey(state.deletedEntries[k]);
+          });
+        } else {
+          Object.keys(state.original).map((k) => {
+            if (filterVal === "untranslated" && state.translated[k]) {
+              return;
+            }
+            if (
+              filterVal === "changed" &&
+              state.changedEntries.indexOf(k) === -1
+            ) {
+              return;
+            }
+            if (
+              filterVal === "deleted" &&
```

Comment:
This `if` block may never run. It is in the `else` branch of `if (filterVal === "deleted")` on line 273.